### ISSUE TITLE
Fail to tokenize a regular expression

### ISSUE THREAD

**Issue by vitaly-t** [jquery/esprima#1873]
Fails to process the following code:

```js
const input = 'if(true){}/\\d/.test(\'\');'
```
or
```js
const input = "if(true){}/\\d/.test('');";
```

This is a perfectly valid JavaScript after compression that Esprima cannot handle - throws an error:

```
Error: Line 1: Unexpected token ILLEGAL
    at ErrorHandler.constructError (D:\NodeJS\tests\node_modules\esprima\dist\esprima.js:5004:22)
    at ErrorHandler.createError (D:\NodeJS\tests\node_modules\esprima\dist\esprima.js:5020:27)
    at ErrorHandler.throwError (D:\NodeJS\tests\node_modules\esprima\dist\esprima.js:5027:21)
    at Scanner.throwUnexpectedToken (D:\NodeJS\tests\node_modules\esprima\dist\esprima.js:5155:35)
    at Scanner.getComplexIdentifier (D:\NodeJS\tests\node_modules\esprima\dist\esprima.js:5488:23)
    at Scanner.scanIdentifier (D:\NodeJS\tests\node_modules\esprima\dist\esprima.js:5556:67)
    at Scanner.lex (D:\NodeJS\tests\node_modules\esprima\dist\esprima.js:6223:26)
    at Tokenizer.getNextToken (D:\NodeJS\tests\node_modules\esprima\dist\esprima.js:6666:84)
    at Object.tokenize (D:\NodeJS\tests\node_modules\esprima\dist\esprima.js:154:36)
```

**Comment by crazyhuggins** [comment]
@vitaly-t A friendly advice. If you report an issue on Esprima repo @mikesherov will claim you are a troll. Even if each of your report follows the GH guidlines. So maybe best not to report anything on this repo? There exist evidence for this, and screenshots have been taken.  There also seem to exist claim that I'm two different users in here from same person, therefor I choose to warn you.  I haven't reported much issues on this repo, but I know the other user have.

**Comment by vitaly-t** [comment]
@ariya Any chance this 100% bug will be looked at? It was logged 18 days ago, and no response since.

**Comment by ariya** [comment]
@vitaly-t Please be patient, every contributor seems to be extremely busy these days.

**Comment by michaelficarra** [comment]
@vitaly-t What version of esprima are you using? I cannot reproduce a failure using the [online demo](http://esprima.org/demo/parse.html?code=if(true)%7B%7D%2Fd%2F.test('')%3B).

**Comment by vitaly-t** [comment]
@michaelficarra your code isn't what I showed, this one is: http://esprima.org/demo/parse.html?code=if(true)%7B%7D%5C%2Fd%2F.test(%27%27)%3B

**Comment by vitaly-t** [comment]
@michaelficarra There is something not right with that online test. I cannot get it to fail, but locally it fails always, any any 3.x or 4.0.0 version of Esprima.

You must define the test correctly, like this:

```js
const input = 'if(true){}/\\d/.test(\'\');';
```

**Comment by vitaly-t** [comment]
Actually no, I can, and [here it is](http://esprima.org/demo/parse.html?code=if(true)%7B%7D%2F%5C%5Cd%2F.test(%5C'%5C')%3B).

**Comment by michaelficarra** [comment]
@vitaly-t That's no a JavaScript program. Check your quotes.

**Comment by vitaly-t** [comment]
@michaelficarra it perfectly is JavaScript. You can change it to this, if you like:

```js
const input = "if(true){}/\\d/.test('');";
```

**Comment by michaelficarra** [comment]
@vitaly-t Sorry, but you're just wrong.

```
$ npm install esprima
+ esprima@4.0.0
added 1 package in 1.318s
$ echo "if(true){}/\\d/.test('');" > test.js
$ node_modules/.bin/esvalidate test.js
$ echo $?
0
```

I cannot reproduce this either on the website or locally using a freshly-installed esprima `4.0.0`. Which do you think is more likely: esprima, a widely used parser, is failing to parse a very simple program, or you are having trouble giving it the correct input?

**Comment by ariya** [comment]
Another way to play with this is with RunKit, e.g. https://runkit.com/5a127690868980001150b6df/5a127690bb5c880012193356.

**Comment by vitaly-t** [comment]
In a simplified form, this is the code that fails:

```js
const esprima = require('esprima');
const input = "if(true){}/\\d/.test('');";
esprima.tokenize(input);
```

**Comment by ariya** [comment]
@vitaly-t Thanks for the clarification! So this is basically a problem for the pure tokenizer (not the parser).

**Comment by vitaly-t** [comment]
@ariya it would appear, yes, sorry for the confusion earlier. The tokenizer fails with the given example.

**Comment by vitaly-t** [comment]
@ariya Just tested against version 4.0.1, and the issue is still there. So whatever that commit was - does not fix the issue. Please re-open it.

i'm getting the very same error:

```
Error: Line 1: Unexpected token ILLEGAL
    at ErrorHandler.constructError (D:\NodeJS\decomment\node_modules\esprima\dist\esprima.js:5012:22)
    at ErrorHandler.createError (D:\NodeJS\decomment\node_modules\esprima\dist\esprima.js:5028:27)
    at ErrorHandler.throwError (D:\NodeJS\decomment\node_modules\esprima\dist\esprima.js:5035:21)
    at Scanner.throwUnexpectedToken (D:\NodeJS\decomment\node_modules\esprima\dist\esprima.js:5164:35)
    at Scanner.getComplexIdentifier (D:\NodeJS\decomment\node_modules\esprima\dist\esprima.js:5497:23)
    at Scanner.scanIdentifier (D:\NodeJS\decomment\node_modules\esprima\dist\esprima.js:5565:67)
    at Scanner.lex (D:\NodeJS\decomment\node_modules\esprima\dist\esprima.js:6232:26)
    at Tokenizer.getNextToken (D:\NodeJS\decomment\node_modules\esprima\dist\esprima.js:6675:84)
    at Object.tokenize (D:\NodeJS\decomment\node_modules\esprima\dist\esprima.js:154:36)
    at Object.parseRegEx (D:\NodeJS\decomment\lib\utils.js:37:13)

```

**Comment by ariya** [comment]
@vitaly-t If you want to try it out, can you try the master branch?
The fix was not included in 4.0.1, it is too risky for a patch release.
